### DISCUSSION TITLE
Very low performance (lower than TFT_eSPI) when using writePixelsDMA

### DISCUSSION THREAD

**Discussion by daywink** [lovyan03/LovyanGFX#381] [Q&A] [answered]
Hello, I am trying to use LVGL with my M5Stack Core2.
Because I see this library easily supports it, I want to use it.
I am unable to get LVGL running at a fast speed. I ported the exact same code, but only getting 16 FPS with LovyanGFX and 24 FPS with TFT_eSPI.

**Comment by lovyan03** [comment]
https://github.com/lovyan03/LovyanGFX/issues/374#issuecomment-1478718157

**Reply by daywink** [reply]
I tested using this code:

```ino
/////////////////////////////////////////////////////////////////
/*
  ESP32 | LVGL8 | Ep 0. GFX Setup (ft. LovyanGFX)
  Video Tutorial: https://youtu.be/IPCvQ4o_WP8
  Created by Eric N. (ThatProject)
*/
/////////////////////////////////////////////////////////////////
#define LGFX_M5STACK_CORE2 
#include <lvgl.h>
#define LGFX_USE_V1
#include <LovyanGFX.hpp>
#include <LGFX_AUTODETECT.hpp>
#include <demos/lv_demos.h>

LGFX tft;

/*Change to your screen resolution*/
static const uint32_t screenWidth  = 320;
static const uint32_t screenHeight = 240;
static lv_disp_draw_buf_t draw_buf;
static lv_color_t  buf[ screenWidth * screenHeight / 10 ];
static lv_color_t buf2[ screenWidth * screenHeight / 10 ];


/* Display flushing */
void my_disp_flush(lv_disp_drv_t *disp, const lv_area_t *area, lv_color_t *color_p){
  tft.pushImageDMA( area->x1
                  , area->y1
                  , area->x2 - area->x1 + 1
                  , area->y2 - area->y1 + 1
                  , ( lgfx::swap565_t* )&color_p->full );
  lv_disp_flush_ready(disp); 
}

/*Read the touchpad*/
void my_touchpad_read( lv_indev_drv_t * indev_driver, lv_indev_data_t * data )
{
   uint16_t touchX, touchY;
   bool touched = tft.getTouch( &touchX, &touchY);
   if( !touched )
   {
      data->state = LV_INDEV_STATE_REL;
   }
   else
   {
      data->state = LV_INDEV_STATE_PR;

      /*Set the coordinates*/
      data->point.x = touchX;
      data->point.y = touchY;
   }
}

void setup()
{
   Serial.begin(115200);

   tft.begin();        
   tft.setRotation(1);
   tft.setBrightness(255);

   lv_init();
   lv_disp_draw_buf_init( &draw_buf, buf, buf2, screenWidth * screenHeight / 10 );

   /*Initialize the display*/
   static lv_disp_drv_t disp_drv;
   lv_disp_drv_init(&disp_drv);

   /*Change the following line to your display resolution*/
   disp_drv.hor_res = screenWidth;
   disp_drv.ver_res = screenHeight;
   disp_drv.flush_cb = my_disp_flush;
   disp_drv.draw_buf = &draw_buf;
   lv_disp_drv_register(&disp_drv);

   /*Initialize the (dummy) input device driver*/
   static lv_indev_drv_t indev_drv;
   lv_indev_drv_init(&indev_drv);
   indev_drv.type = LV_INDEV_TYPE_POINTER;
   indev_drv.read_cb = my_touchpad_read;
   lv_indev_drv_register(&indev_drv);

 lv_demo_music();
}

void loop()
{
   lv_timer_handler(); /* let the GUI do its work */

}
```

15 FPS. Lower performance than last code

**Reply by tobozo** [reply]
does it improve when you disable psram ?

**Reply by daywink** [reply]
PSRAM disabled, performance is 17 FPS. Still very low.

**Reply by tobozo** [reply]
can you share your lvgl settings so I can to reproduce that with my M5core2?

It would probably make no difference, but what happens if you change this:

```cpp
#define LGFX_USE_V1
#include <LovyanGFX.hpp>
#include <LGFX_AUTODETECT.hpp>
```

by this:

```cpp
#define LGFX_USE_V1
#define LGFX_AUTODETECT
#include <LovyanGFX.hpp>
```

**Reply by daywink** [reply]
The LVGL settings are all set to the default, except for setting the colour format to 16 swap and also enabling LV_TICK_CUSTOM.
What you just said makes no visible difference to me, the only difference is maybe worse performance by 1FPS, but I am not 100% sure about it.
My LVGL settings are the same as `lv_conf_template.h`, except for:

```c
#define LV_COLOR_16_SWAP 1
#define LV_COLOR_SCREEN_TRANSP 1
#define LV_COLOR_MIX_ROUND_OFS 128
#define LV_TICK_CUSTOM 1
#define LV_FONT_MONTSERRAT_12 1
#define LV_FONT_MONTSERRAT_14 1
#define LV_FONT_MONTSERRAT_16 1
#define LV_USE_DEMO_MUSIC 1
#define LV_DEMO_MUSIC_SQUARE    1
#define LV_DEMO_MUSIC_AUTO_PLAY 1
```

I forgot to mention that my performance goal is this: https://www.youtube.com/watch?v=Zq_MZXYbdgI

**Reply by lovyan03** [reply]
```
void setup()
{
   Serial.begin(115200);

   tft.begin();        
   tft.startWrite();  // please add this line.
   tft.setRotation(1);
```

**Comment by tobozo** [comment]
Do you have a platformio.ini file to share, if not could you share the lvgl version and installation steps you took to get your music demo working?

I managed to get [m5core2-lvgl-demo](https://github.com/mhaberler/m5core2-lvgl-demo) to work with M5Unified but it's based on `lv_arduino@^3.0.1` which appears to be archived, and doesn't include a working music demo I could compare with your results.

**Reply by daywink** [reply]
I am just using Arduino IDE, but the LVGL version is the latest release which you can find [here](https://github.com/lvgl/lvgl/releases/tag/v8.3.5).
Your version of LVGL is very old and support for it has long since dropped.

**Reply by tobozo** [reply]
I managed to get lvgl 8.x to work, and got the music demo to show an average of 25fps by playing with the config (setting `LV_DISP_DEF_REFR_PERIOD` to 10, disabling psram, smaller buffers)

**Reply by daywink** [reply]
This seems like a "trick" for a better average FPS value, if you can, can you please record a video of the music demo?
Because when sitting still, it can render very fast as there is no difference from the previous frame, which gives a boost in average FPS, but doesn't make it much smoother.
Also, I made a video comparing TFT_eSPI and LovyanGFX with the same demo.
https://www.youtube.com/watch?v=rIlqNHV7Jok&feature=youtu.be
I am sorry for the camera shake in the video.

**Reply by tobozo** [reply]
Comparing TFT_eSPI and LovyanGFX between two different devices should be done both ways for this test to be fully conclusive.

My M5Core2 shows the same average speed as yours when I use your lv_conf.

However TFT_eSPI does not support M5Core2 and I could not manage to test it [using this config](https://community.m5stack.com/topic/4278/stand-alone-tft_espi-library-on-core2/2) and got the exact same results.

Performing the lvgl test with LovyanGFX on your **other device** could confirm if the problem is isolated to LovyanGFX or just the M5Core2.

**Reply by daywink** [reply]
Using LovyanGFX on the M5Core2 with the external display (one on the right) yields the *exact* same performance as the M5Core2 with integrated display running LovyanGFX.

Just incase the message is too confusing:
Both devices in the video are M5Core2, just one is only the main board with a battery and an external display. I am using my own AXP192 library to manage everything.

Here is another comparison video I made, in here even LovyanGFX is having some problems and making artefacts on the display.
https://www.youtube.com/watch?v=IyRWrzWhHSQ

**Reply by lovyan03** [reply]
The first thing to understand is that according to the ILI9342C spec sheet, operation is guaranteed up to 10 MHz.
In my experience, it works stably up to 40MHz, but you should understand that this is a setting outside the guaranteed operation.
Communication at 80MHz will most certainly cause glitches.

![image](https://user-images.githubusercontent.com/42724151/229250073-973cd512-0ea3-474d-9fad-9d356fe37e44.png)

**Reply by daywink** [reply]
I am aware about this. I am running the M5Core2 display at 40 MHz, the same speed I am running my external display (ILI9341) at.

**Reply by lovyan03** [reply]
![image](https://user-images.githubusercontent.com/42724151/229250865-2b56c084-df33-4e45-9a73-828f42141c5b.png)

ok. It is only necessary to understand that the artifacts are not a problem with LovyanGFX, but with the configuration.

**Reply by daywink** [reply]
TFT_eSPI and LovyanGFX were both set to run at 80 MHz in that video. My display can handle up to 115 MHz before it starts missing clock pulses and corrupting (tested with Teensy 4.1 at 115 MHz, no artifacts).

Here is the LGFX configuration:
```cpp
class LGFX : public lgfx::LGFX_Device
{
  lgfx::Panel_ILI9341     _panel_instance;

  lgfx::Bus_SPI        _bus_instance;   // SPI връзка


public:

// Създайте конструктор и задайте различни настройки тук.
// Ако промените името на класа, задайте същото име за конструктора.
  LGFX(void)
  {
    { // Конфигурирайте настройките за управление на буса
      auto cfg = _bus_instance.config();    // Вземете структура за настройки на буса

// SPI настройки
      cfg.spi_host = VSPI_HOST;     // Изберете SPI, за да използвате ESP32-S2,C3: SPI2_HOST или SPI3_HOST / ESP32: VSPI_HOST или HSPI_HOST
      // » С надграждането на версията на ESP-IDF, описанието на VSPI_HOST и HSPI_HOST ще бъде отхвърлено, така че ако възникне грешка, използвайте SPI2_HOST и SPI3_HOST вместо това.
      cfg.spi_mode = 0;             // SPI режим (0 ~ 3)
      cfg.freq_write = 80000000;    // SPI тактова честота за предаване (до 80MHz, закръглена до стойността, получена чрез разделяне на 80MHz на цяло число)
      cfg.freq_read  = 16000000;    // SPI тактова честота за получаване
      cfg.spi_3wire  = true;        // Задайте true, ако приемането се извършва на MOSI pin
      cfg.use_lock   = true;        // Задайте true, за да използвате заключване на транзакция
      cfg.dma_channel = SPI_DMA_CH_AUTO; // Задайте DMA канал за използване (0=да не се използва DMA / 1=1ch / 2=ch / SPI_DMA_CH_AUTO=автоматична настройка)
      // » С надграждане на версията на ESP-IDF, SPI_DMA_CH_AUTO (автоматична настройка) се препоръчва за DMA канала. Посочването на 1ch и 2ch е отхвърлено.
      cfg.pin_sclk = 18;            // Pin номер, където е свързан SCLK (нужно)
      cfg.pin_mosi = 23;            // Pin номер, където е свързан MOSI (нужно)
      cfg.pin_miso = 19;            // Pin номер, където е свързан MISO (-1 = изкл.)
      cfg.pin_dc   = 27;            // Pin номер, където е свързан DC   (-1 = изкл.)

      _bus_instance.config(cfg);    // Прилага зададената стойност към шината
      _panel_instance.setBus(&_bus_instance);      // Задайте буса на панела
    }

    { // Задайте управлението на панела на дисплея
      auto cfg = _panel_instance.config();    // Получава структурата за настройките на панела на дисплея

      cfg.pin_cs           =    14;  // Pin номер, където е свързан CS   (-1 = изкл.)
      cfg.pin_rst          =    33;  // Pin номер, където е свързан RST  (-1 = изкл.)
      cfg.pin_busy         =    -1;  // Pin номер, където е свързан BUSY (-1 = изкл.)

      // » Следните стойности на настройката са общи начални стойности за всеки панел, така че, моля, коментирайте всички непознати елементи и ги опитайте.

      cfg.panel_width      =   240;  // Реална видима ширина
      cfg.panel_height     =   320;  // Реална видима височина
      cfg.offset_x         =     0;  // Стойност на изместване на панела в посока X
      cfg.offset_y         =     0;  // Стойност на изместване на панела в посока Y
      cfg.offset_rotation  =     0;  // Отместване на стойността на посоката на въртене 0-7 (4-7 е наобратно)
      cfg.dummy_read_pixel =     8;  // Брой битове за фиктивно четене преди четене на пиксел
      cfg.dummy_read_bits  =     1;  // Брой битове за фиктивно четене преди четене на непикселни данни
      cfg.readable         =  true;  // Задайте true, ако данните могат да бъдат прочетени
      cfg.invert           = false;  // Задайте true, ако светлината/тъмнината на панела е обърната
      cfg.rgb_order        = false;  // Задайте true, ако червеното и синьото на панела са разменени
      cfg.dlen_16bit       = false;  // Задайте true за панели, които предават дължина на данни в 16-битови единици с 16-битов паралел или SPI
      cfg.bus_shared       =  true;  // Ако шината е споделена със SD картата, задайте true (управление на шината с drawJpgFile и т.н.)

// Задайте следното само когато дисплеят се измества с драйвер с променлив брой пиксели, като например ST7735 или ILI9163.
//    cfg.memory_width     =   240;  // Максимална ширина, поддържана от драйвера IC
//    cfg.memory_height    =   320;  // Максимална височина, поддържана от драйвера IC

      _panel_instance.config(cfg);
    }
    setPanel(&_panel_instance); // Настройва панела за използване
  }
};
```


(edit)
I don't know how this can help, but here are the TFT_eSPI configurations (pin numbers are different)
```cpp
#define ILI9341_DRIVER

                    // Typical board default pins
#define TFT_CS   27 //     10 or 34

#define TFT_MOSI 23 //     11 or 35
#define TFT_SCLK 18 //     12 or 36
#define TFT_MISO 38 //     13 or 37

#define TFT_DC   13
#define TFT_RST  19

//#define TOUCH_CS 16 // Optional for touch screen
#define TOUCH_CS -1 // disable warning

#define LOAD_GLCD
#define LOAD_FONT2
#define LOAD_FONT4
#define LOAD_FONT6
#define LOAD_FONT7
#define LOAD_FONT8
#define LOAD_GFXFF

#define SMOOTH_FONT

// FSPI port will be used unless the following is defined
#define USE_VSPI_PORT

//#define SPI_FREQUENCY  27000000
#define SPI_FREQUENCY  80000000   // Maximum for ESP32

#define SPI_READ_FREQUENCY  6000000

#define SPI_TOUCH_FREQUENCY 2500000
```

**Comment by lovyan03** [comment]
https://twitter.com/lovyan03/status/1504276217339727873

**Reply by daywink** [reply]
Looks interesting. Can you share the code?

**Reply by lovyan03** [reply]
https://github.com/lovyan03/LovyanGFX/tree/develop/examples/Advanced/LVGL_PlatformIO

**Reply by tobozo** [reply]
@lovyan03 thanks for that example, is it worth creating a [github template project](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository) out of this?

@windorey it seems like the missing call to `tft.startWrite();` inside `flush_cb` was the curlpit, I'm getting a better refresh rate in every test since I added it.

**Reply by daywink** [reply]
Frame rate is still not very good.
18 FPS on Music demo. TFT_eSPI gets 20, and it's significantly smoother. Is there anything I can do?

**Reply by tobozo** [reply, marked as answer]
there's something going on with M5Core2 board settings as I had significant fps improvements using an esp32-dev profile instead, in this profile the settings are inherited from the wroom and psram is *really* disabled. 

![image](https://user-images.githubusercontent.com/1893754/229600592-676ed737-6815-421c-9900-3d496352dabe.png)

https://user-images.githubusercontent.com/1893754/229605270-893654f8-d750-4654-9d18-fa8c26ab1fc2.mp4

**Reply by daywink** [reply]
This is very interesting, I'll test out on my "Core2" and also on my Core2 later. (One of the Core2's is in quotation marks because it's a Core2 motherboard with external screen.)

**Reply by daywink** [reply]
https://youtu.be/7weNrjTr_xU

I am getting the same results, thank you for this heads-up!
